### PR TITLE
Added CUDA 11.0 to CI

### PR DESCRIPTION
**Pull request by tgaddair**
Adds TF 2.4.0, PyTorch 1.7.1, MXNet nightly with CUDA 11.0 tests.

Also enforces that tests in the GPU Buildkite queues have CUDA / GPUs available. Previously, TensorFlow was compiling successfully but not enabling GPU support with CUDA 11.0.

### CODE REVIEW

File: .buildkite/gen-pipeline.sh [hunk 1]
```diff
@@ -11,21 +11,21 @@ dir="$(dirname "$0")"
...
-  printf "test-cpu-gloo-py3_6-tf1_15_0-keras2_2_4-torch1_2_0-mxnet1_4_1-pyspark2_3_2 "
-  printf "test-cpu-openmpi-py3_6-tf2_0_0-keras2_2_4-torch1_3_0-mxnet1_4_1-pyspark2_4_7 "
-  printf "test-cpu-openmpi-py3_6-tf2_1_0-keras2_3_1-torch1_4_0-mxnet1_5_0-pyspark2_4_7 "
-  printf "test-cpu-gloo-py3_7-tf2_2_0-keras2_3_1-torch1_5_0-mxnet1_5_0-pyspark2_4_7 "
-  printf "test-cpu-gloo-py3_8-tf2_3_0-keras2_3_1-torch1_6_0-mxnet1_5_0-pyspark3_0_1 "
+  printf "test-cpu-gloo-py3_6-tf2_0_0-keras2_2_4-torch1_3_0-mxnet1_4_1-pyspark2_3_2 "
+  printf "test-cpu-openmpi-py3_6-tf2_1_0-keras2_2_4-torch1_4_0-mxnet1_4_1-pyspark2_4_7 "
+  printf "test-cpu-openmpi-py3_6-tf2_2_0-keras2_3_1-torch1_5_0-mxnet1_5_0-pyspark2_4_7 "
+  printf "test-cpu-gloo-py3_7-tf2_3_1-keras2_3_1-torch1_6_0-mxnet1_5_0-pyspark2_4_7 "
+  printf "test-cpu-gloo-py3_8-tf2_4_0-keras2_3_1-torch1_7_1-mxnet1_5_0-pyspark3_0_1 "
```
Comment (EnricoMi) [inline .buildkite/gen-pipeline.sh:18]: to summarize test environment changes:
- we promote tf versions for cpu-gloo in their respective py version by one version: tf1_15 -> tf2_0, tf2_0 -> tf2_1, tf2_3 -> tf2_4
- we do the same promotion for cpu-openmpi py versions
- torch promotes accordingly to their tf version

File: .buildkite/gen-pipeline.sh [hunk 2]
```diff
@@ -11,21 +11,21 @@ dir="$(dirname "$0")"
...
-   printf "test-gpu-openmpi-gloo-py3_6-tf2_2_0-keras2_3_1-torch1_5_0-mxnet1_4_1-pyspark2_4_7 "
-   printf "test-gpu-openmpi-py3_6-tf2_3_0-keras2_3_1-torch1_6_0-mxnet1_6_0-pyspark2_4_7 "
-   printf "test-gpu-openmpi-py3_6-tfhead-kerashead-torchhead-mxnethead-pyspark2_4_7 "
-   printf "test-mixed-openmpi-py3_6-tf1_15_0-keras2_3_1-torch1_4_0-mxnet1_5_0-pyspark2_4_7 "
+  printf "test-gpu-openmpi-py3_6-tf1_15_0-keras2_2_4-torch1_3_0-mxnet1_4_1-pyspark2_4_7 "
+  printf "test-gpu-gloo-py3_6-tf2_0_0-keras2_3_1-torch1_4_0-mxnet1_4_1-pyspark2_4_7 "
+  printf "test-gpu-openmpi-py3_6-tf2_3_1-keras2_3_1-torch1_6_0-mxnet1_6_0-pyspark2_4_7 "
+  printf "test-gpu-openmpi-gloo-py3_6-tf2_4_0-keras2_3_1-torch1_7_1-mxnethead-pyspark2_4_7 "
+  printf "test-gpu-openmpi-py3_6-tfhead-kerashead-torchhead-mxnethead-pyspark2_4_7 "
+  printf "test-mixed-openmpi-py3_6-tf1_15_0-keras2_3_1-torch1_4_0-mxnet1_5_0-pyspark2_4_7 "
```
Comment (EnricoMi) [inline .buildkite/gen-pipeline.sh:28]: and here we
- move gpu-openmpi to the next tf patch version
- move gpu-openmpi-gloo to the next but one minor tf version to test msnethead

File: .buildkite/gen-pipeline.sh [hunk 3]
```diff
@@ -11,21 +11,21 @@ dir="$(dirname "$0")"
...
 fi)
```
Comment (EnricoMi) [inline .buildkite/gen-pipeline.sh:29]: I find all these combinations very confusion and hard to follow which tf we combine with which keras and torch and mxnet. How do you feel about it?

File: docker-compose.test.yml [hunk 4]
```diff
@@ -195,43 +195,43 @@ services:
...
+        CUDNN_VERSION: 8.0.5.39-1+cuda11.0
+        NCCL_VERSION_OVERRIDE: 2.7.8-1+cuda11.0
         MPI_KIND: OpenMPI
         PYTHON_VERSION: 3.6
         TENSORFLOW_PACKAGE: tf-nightly-gpu
         KERAS_PACKAGE: keras==2.4.3
         PYTORCH_PACKAGE: torch-nightly
         TORCHVISION_PACKAGE: torchvision
-        MXNET_PACKAGE: mxnet-nightly
+        MXNET_PACKAGE: mxnet-cu110
```
Comment (EnricoMi) [inline docker-compose.test.yml:234]: is this still a head version of mxnet?

File: test/parallel/test_mxnet.py [hunk 5]
```diff
@@ -52,6 +62,10 @@ def filter_supported_types(self, types):
            types = [t for t in types if t in ccl_supported_types]
         return types
 
+    def test_gpu_required(self):
```
Comment (EnricoMi) [inline test/parallel/test_mxnet.py:65]: this method does not seem to be used anywhere, same for the same-name-methods in two other files

**Review by EnricoMi — COMMENTED** [review]
Is updating the eigen submodule expected to be part of this PR?

Such a change always breaks my git repo and I cannot apply such a change in master to my local master. Are there any instructions how to do this properly, i.e. how to get my submodule upgraded to the new commit?

**Comment by EnricoMi** [comment]
Unit test results say there are 33 less unit tests, is this expected?

**Comment by EnricoMi** [comment]
These tests went missing:

```
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_auto_scale_down_by_discovery
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_auto_scale_down_by_exception
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_auto_scale_no_spark_black_list
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_auto_scale_spark_blacklist_no_executor_reuse
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_auto_scale_spark_blacklist_no_executor_reuse_in_app
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_auto_scale_spark_blacklist_no_executor_reuse_same_task
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_auto_scale_spark_blacklist_no_node_reuse
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_auto_scale_spark_blacklist_no_node_reuse_in_app
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_auto_scale_up
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_fault_tolerance_all_hosts_lost
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_fault_tolerance_exception_all_ranks
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_fault_tolerance_exception_single_rank
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_fault_tolerance_exception_with_min_hosts_timeout
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_fault_tolerance_hosts_added_and_removed
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_fault_tolerance_no_spark_blacklist
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_fault_tolerance_spark_blacklist_no_executor_reuse
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_fault_tolerance_spark_blacklist_no_executor_reuse_in_app
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_fault_tolerance_spark_blacklist_no_executor_reuse_same_task
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_fault_tolerance_spark_blacklist_no_node_reuse
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_fault_tolerance_spark_blacklist_no_node_reuse_in_app
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_fault_tolerance_unused_hosts_added_and_removed
test.integration.test_elastic_spark_tensorflow.ElasticSparkTensorflowTests:test_happy_run
test.integration.test_elastic_tensorflow.ElasticTensorFlowTests:test_all_hosts_blacklisted
test.integration.test_elastic_tensorflow.ElasticTensorFlowTests:test_all_ranks_failure
test.integration.test_elastic_tensorflow.ElasticTensorFlowTests:test_fault_tolerance_without_scaling
test.integration.test_elastic_tensorflow.ElasticTensorFlowTests:test_hosts_added_and_removed
test.integration.test_elastic_tensorflow.ElasticTensorFlowTests:test_min_hosts_timeout
test.integration.test_elastic_tensorflow.ElasticTensorFlowTests:test_reset_limit
test.integration.test_elastic_tensorflow.ElasticTensorFlowTests:test_single_rank_failure
test.integration.test_elastic_tensorflow_keras.ElasticTensorFlowKerasTests:test_all_hosts_blacklisted
test.integration.test_elastic_tensorflow_keras.ElasticTensorFlowKerasTests:test_all_ranks_failure
test.integration.test_elastic_tensorflow_keras.ElasticTensorFlowKerasTests:test_fault_tolerance_without_scaling
test.integration.test_elastic_tensorflow_keras.ElasticTensorFlowKerasTests:test_hosts_added_and_removed
test.integration.test_elastic_tensorflow_keras.ElasticTensorFlowKerasTests:test_min_hosts_timeout
test.integration.test_elastic_tensorflow_keras.ElasticTensorFlowKerasTests:test_reset_limit
test.integration.test_elastic_tensorflow_keras.ElasticTensorFlowKerasTests:test_single_rank_failure
```
Looks like all tensorflow 1.15 spark integration tests.

This one is new (this answers my question where this is called, it is a test):
```
test.parallel.test_mxnet.MXTests:test_gpu_required
```

File: .buildkite/gen-pipeline.sh (hunk 1 above)
Comment (tgaddair) [reply .buildkite/gen-pipeline.sh:18]: Yes, basically making room for tf 2.4.0 and pytorch 1.7.1

File: .buildkite/gen-pipeline.sh (hunk 3 above)
Comment (tgaddair) [reply .buildkite/gen-pipeline.sh:29]: I agree, it's pretty hard to read.  Not sure what a better solution looks like at the moment.  I think dropping keras would make sense, however, as it's basically deprecated.

File: docker-compose.test.yml (hunk 4 above)
Comment (tgaddair) [reply docker-compose.test.yml:234]: It is, yeah.  Possibly we could change this `mxnet-nightly-cu110` if it's more clear.

File: test/parallel/test_mxnet.py (hunk 5 above)
Comment (tgaddair) [reply test/parallel/test_mxnet.py:65]: The way these tests work is that if `HOROVOD_TEST_GPU` is set, then we require that the framework can use GPUs.  This is to workaround an issue where TensorFlow was compiling with GPU support, but unable to use it because the wrong version of CUDA was installed.

**Comment by tgaddair** [comment]
@EnricoMi changed the TensorFlow 1.15 test to be both MPI and Gloo, so we should get those tests back.  Good catch!

**Comment by EnricoMi** [comment]
> @EnricoMi changed the TensorFlow 1.15 test to be both MPI and Gloo, so we should get those tests back. Good catch!

Thanks to the unit test action :smile:.

**Comment by EnricoMi** [comment]
Unit test results confirm those spark integration tests are back (537).

     537 tests +    3     507 :heavy_check_mark: +    2       30 :zzz: +  1  0 :x: ±0 

They also confirm those three `test_gpu_required` tests have been added (+3) but also that one of them is always skipped (:zzz: +1), so one of them never runs with `HOROVOD_TEST_GPU=1`.

**Comment by tgaddair** [comment]
> Unit test results confirm those spark integration tests are back (537).
> 
>      537 tests +    3     507 ✔️ +    2       30 💤 +  1  0 ❌ ±0 
> 
> They also confirm those three `test_gpu_required` tests have been added (+3) but also that one of them is always skipped (💤 +1), so one of them never runs with `HOROVOD_TEST_GPU=1`.

That makes sense, as the added tests were CPU-based, so the `test_gpu_required` test should be skipped.

**Comment by EnricoMi** [comment]
> That makes sense, as the added tests were CPU-based, so the `test_gpu_required` test should be skipped.

A skipped test means across all its runs the test is always skipped, it never passes. In the context of these three tests it means it never had `has_gpu == True` and never ran with `HOROVOD_TEST_GPU=1`, but all three tests are in `test/parallel` and that is tested with GPU.

Anyway, I scanned the test result files and could find runs that are not skipped for all of the three tests. I will look into the unit test action to have it report on the skipped tests, then we know which tests are considered as always skipped and if this is true.

**Review by EnricoMi — APPROVED** [review]
LGTM!

**Comment by EnricoMi** [comment]
Interestingly, stats on merge commit are now what I would have expected:

     537 tests +       3     508 :heavy_check_mark: +    3       29 :zzz: ±    0  0 :x: ±0
